### PR TITLE
Ranged upgrades

### PR DESCRIPTION
**Pull request by kattalainen**
Made two upgrades for range. First one is a life steal upgrade that adds health to the player based on damage done with ranged attacks. The other increases the projectile velocity. The velocity upgrade has a max level that caps velocity at 100 due to collision detection limitations.
Pressing "u" will apply all available upgrades to the player. 
All values such as prices, velocity and life steal percentage can and probably will be changed as needed in the future.

### CODE REVIEW

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeRangeLifeSteal.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+{
+	// get damage from projectile
+	int damage = m_pParentEntity->GetComponent<component::ProjectileComponent>()->GetDamage();
+	// calculate amount of health to add
+	int hpChange = static_cast<int>(damage * m_PercentageGain);
+	// add a bit of health
+	Player::GetInstance().GetPlayer()->GetComponent<component::HealthComponent>()->ChangeHealth(hpChange);
+
+	// increase price for next level
+	m_Price *= 2;
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeRangeLifeSteal.cpp:47]: Increasing price of next level for every hit? Sounds expensive!

**Review by Schtekt — DISMISSED** [review]
Looks good, just a minor tweak that needs fixing :)

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeRangeLifeSteal.cpp (hunk 1 above)
Comment (kattalainen) [reply StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeRangeLifeSteal.cpp:47]: oops :> fixed it!

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h [hunk 2]
```diff
@@ -27,8 +27,13 @@ namespace component
 
 		int GetHealth();
 
+		int GetMaxHealth() const;
```
Comment (fredrix96) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h:30]: const int

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h [hunk 3]
```diff
@@ -27,8 +27,13 @@ namespace component
 
 		int GetHealth();
```
Comment (fredrix96) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h]: const int

Comment (fredrix96) [reply StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h]: () const;

**Review by fredrix96 — DISMISSED** [review]
Seems to work fine and the code standard is good

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.cpp [hunk 4]
```diff
@@ -53,12 +53,12 @@ void component::HealthComponent::ChangeHealth(int hpChange)
 
 }
 
-int component::HealthComponent::GetHealth()
+const int component::HealthComponent::GetHealth() const
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.cpp]: Remove the first const keyword. We do want to be able to use the variable that we get from this.
As it is not a pointer the user will NOT be able to change the value in the class.
"int component::HealthComponent::GetHealth() const"

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.cpp [hunk 5]
```diff
@@ -53,12 +53,12 @@ void component::HealthComponent::ChangeHealth(int hpChange)
...
 {
 	return m_Health;
 }
 
-int component::HealthComponent::GetMaxHealth() const
+const int component::HealthComponent::GetMaxHealth() const
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.cpp]: Remove first const keyword.

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h [hunk 6]
```diff
@@ -25,9 +25,9 @@ namespace component
 		// will publish a death event if hp reaches 0 or bellow
 		void ChangeHealth(int hpChange);
 
-		int GetHealth();
+		const int GetHealth() const;
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h]: Remove first const keyword.

File: StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h [hunk 7]
```diff
@@ -25,9 +25,9 @@ namespace component
...
 
-		int GetMaxHealth() const;
+		const int GetMaxHealth() const;
```
Comment (Schtekt) [inline StortSpelProjekt/Game/src/Gamefiles/Components/HealthComponent.h]: Remove first const keyword.

**Review by Schtekt — DISMISSED** [review]
Adding a const keyword on the lefthand side of a function stops the user from making changes to the given value.
As the value is a copy of the variable we want to provide the user SHOULD be able to change this value as they see fit, it will not change the value of the object where the function is called from.

**Review by Schtekt — APPROVED** [review]

**Review by fredrix96 — APPROVED** [review]
Looks good

**Pull request merged**